### ISSUE TITLE
Support updating existing and creating new datasets

### ISSUE THREAD

**Issue by marks** [Chicago/RSocrata#58]
I have heard this request a few times from Socrata customers and employees

**Comment by marks** [comment]
A snippet of code from @peteraustinmoore (thank you!) to update an existing dataset (HTTP PUT)

```R
library(httr) # Upload to Socrata
d_311 <- read.csv("//.csv")

PUT("https://data.somervillema.gov/resource/vqwi-n3ah.json",
    body = d_311,
    authenticate("#EMAIL", "#Password"),
    add_headers("X-App-Token" = " #App token ",
                "Content-Type" = "text/csv"))
```

**Comment by tomschenkjr** [comment]
Interesting. I think this could be useful

**Comment by stuagano** [comment]
Use Case: Download dataset 
read.socrata("https://data.austintexas.gov/resource/bptg-ndvw.json") 
Create Time Series Variable: 
myts <- ts(EMSIncidents$count_responses_all, start=c(2010, 1), end=c(2015,2), frequency=12)
Forecast:
fit <- stl(myts, s.window = "period")
forecast(fit)
ForecastDate <- data.frame(forecast(fit)) write forecasted data to CSV for storing and eventual port back into dataset.  Does the forecast package return all values or just forecasted values?

**Comment by marks** [comment]
first step complete: being able to query a private dataset

`read.socrata("https://domain.com/resource/abcd-1234.csv",NULL,"mark.silverberg@socrata.com","MY_SECRET_NOT_YOURS")`

TODO
- [ ] write tests
- [ ] confirm I didnt break anything else

**Comment by dmpe** [comment]
@marks you used - and you can correct me - the master branch, right ? Well, RSocrata development uses "dev" branch as master. So master branch is stable, whereas the dev branch is what is the development branch. 

see also #56 my pull request, which might (depending on if it will be merged) add some additional things to consider

**Comment by marks** [comment]
Whoops- so I will cause havoc if I submit a PR based on the master branch? Luckily I'm not far along and can recreate my dev branch from you all's. Sound like a plan?


--
Mark Silverberg
(m) 512 826 7004
http://twitter.com/skram

**Comment by dmpe** [comment]
Most probably yes. Do not do this yet, please. Until at least @tomschenkjr will say if #56 will get merged or not. 
In either case, you should branch-off from `dev` not master. BTW: it would be also nice if dev branch can be displayed as default on Rsocrata (this can be done in the settings of the repository)

**Comment by marks** [comment]
Alrighty, I will be working on this feature request at https://github.com/marks/RSocrata/tree/58-support-update-and-create-datasets

**Comment by marks** [comment]
first step complete: being able to query a private dataset


- [x] be able to read private datasets via
    ```r
read.socrata("https://domain.com/resource/abcd-1234.csv",NULL,"name@domain.com","PASSWORD")
```

- [ ] write tests
- [ ] confirm I didnt break anything else

**Comment by marks** [comment]
@stuagano I made some progress on updating datasets.. here's an example: 

```r

# Get started 
setwd("~/Code/RSocrata")
library(devtools)
dev_mode(on=T)
load_all()

# read dataset
dataframe <- read.socrata("https://opendata.demo.socrata.com/resource/grzz-r3zb.json",NULL,email,password)

# do something to the data
# but first we need to convert the `count_responses_all` column to numeric
dataframe$count_responses_all <- as.numeric(as.character(dataframe$count_responses_all))
# multiply all the count responses by 2
dataframe$count_responses_all <- dataframe$count_responses_all * 2

# write to dataset, replacing all existing rows
write.socrata(dataframe,"https://opendata.demo.socrata.com/resource/grzz-r3zb.json","REPLACE",email,password)

# write to dataset, upserting (appends if no unique row ID is set)
write.socrata(dataframe,"https://opendata.demo.socrata.com/resource/grzz-r3zb.json","UPSERT",email,password)


```

**Comment by marks** [comment]
@dmpe @tomschenkjr - I have completed a first pass at being able to update data from R. I'm trying to think how best to write tests for this considering one would need a username and password to test. Would you be OK with an email and associated password being stored in the repo? How would you suggest I add tests for this functionality?

**Comment by dmpe** [comment]
Not sure if you would add it later, but here https://github.com/marks/RSocrata/blob/58-support-update-and-create-datasets/R/writeData.R#L53 there is need for #' @export...

Recently I found this https://github.com/jennybc/googlesheets/blob/master/internal-projects/10_token-encryption.Rmd. Yes, it uses `.httr-oauth` but for you just create a list of username and password and store it in rds. 
https://github.com/jennybc/googlesheets/blob/master/.travis.yml for later use

It works, but in my recent project I was not able to make it working on travis, somehow.

Then there is https://github.com/hadley/secure, well try it yourself - for me it was unpractical. 
These are solutions I would try, but frankly I dont care..... @tomschenkjr will need to decide.

**Comment by tomschenkjr** [comment]
Catching-up on this. @marks -- great suggestion on the private repo and the ability to upload. R-savvy shops could create ETLs using R and the private data access could be used to use Socrata for "internal" projects.

> Whoops- so I will cause havoc if I submit a PR based on the master branch? Luckily I'm not far along and can recreate my dev branch from you all's. Sound like a plan?


@dmpe is right--and thanks for jumping in--we prefer to submit to the `dev` branch. It's the alpha branch whilst the `master` branch is the stable beta. We submit to CRAN quarterly as the "production" version. `master` is _effectively_ production quality, so we like the first point of integration to be `dev`.

> I have completed a first pass at being able to update data from R. I'm trying to think how best to write tests for this considering one would need a username and password to test. Would you be OK with an email and associated password being stored in the repo? How would you suggest I add tests for this functionality?

I have no problem with email and password in the repo -- but obviously a limit account. We had to make the similar choice for tokens as well. @dmpe has some interesting solutions, but seems heavy if you can create a fake account in the short term. A more sophisticated approach would be great.

Do you have a sense of how other projects that integrate with private datasets/keys handle this issue with Socrata?

**Comment by marks** [comment]
@tomschenkjr thanks for the encouragement... confirming next steps:

- [ ] write tests for this functionality which uses a demo account at Socrata - something like mark.silverberg+rsocrata@socrata.com 
- [ ] submit pull request

**Comment by tomschenkjr** [comment]
Sounds about right.


Tom Schenk Jr.

Chief Data Officer

Department of Innovation and Technology

City of Chicago

(312) 744-2770

tom.schenk@cityofchicago.org

data.cityofchicago.org


________________________________

**Comment by marks** [comment]
@tomschenkjr @dmpe - would you have a chance anytime soon to take a look at the tests I wrote and also https://travis-ci.org/marks/RSocrata/builds/88471261 and tell me if it looks good or if some of those warnings are a cause for concern?

Am honestly a R extension newbie. Thanks!

**Comment by dmpe** [comment]
please oppen a Pull requestt. thanks

**Comment by marks** [comment]
Done: https://github.com/Chicago/RSocrata/pull/63

**Comment by tomschenkjr** [comment]
Closed with 5f92c1f7189f6216f55c5644b5d47499a0031326
